### DISCUSSION TITLE
google authentication with supabase dart package

### DISCUSSION THREAD

**Discussion by sumchans** [supabase/supabase#1885] [Questions] [answered]
I am trying out the supabase authentication with Flutter. Below is the code I have on using the google authentication with supabase. 

final supabaseClient = SupabaseClient('supabase_url','supabase_key');
Future<void> signUp() async {
    final response = await supabaseClient.auth.signUp({
        provider: 'google'
    })
}

the above code gives me two errors even before running it -
1. Getter not found: 'provider'
2. Too few positional arguments: 2 required, 1 given

Please advise!

Thanks all!

**Comment by sumchans** [comment]
Anybody from the Supabase community who can help me out on this one?

**Comment by inian** [comment, marked as answer]
Hi @sumchans, [here](https://github.com/supabase/gotrue-dart/blob/main/test/provider_test.dart) are some examples on how the functions are used. 

[Here](https://github.com/supabase/gotrue-dart/blob/d221fb544451fbd34ae51952b9696df2d3880ff2/test/provider_test.dart#L24) is how you signin with google. Does this help?

**Reply by lpylpyleo** [reply]
Hi @inian , I am new to Supabase and also have problems with oAuth. I noticed that in the [test](https://github.com/supabase/gotrue-dart/blob/d221fb544451fbd34ae51952b9696df2d3880ff2/test/provider_test.dart#L34), we have to concat `access_token` and others to the callback url. So does it mean that we need to have an account in advance?

**Reply by phamhieu** [reply]
@lpylpyleo no you don't need. The callback url should already included these info.

**Reply by inian** [reply]
Hey @lpylpyleo, not sure if I get your question completely. That test is for our OAuth providers and there do we assume that you already have an account with Google when you are signing in with Google. 

You don't need to worry about that particular test since that will be implemented by the OAuth provider you choose.

**Reply by lpylpyleo** [reply]
@inian @phamhieu Thanks for your information.

**Reply by sumchans** [reply]
Hello @inian, thanks for those examples, but still using oauth with Google as the provider still not working for me. I looked at the code and how it is used. I tried to copy it over exactly, but I still see errors.
Here is my function that does the signup with Google -
```
Future<void> signUp() async {
     final response = await supabaseClient.auth.signUp(provider: Provider.google);
}
```
When I hover over the error, it says **2 expected arguments, only one found**.

**Reply by inian** [reply]
Try using `supabaseClient.auth.signIn`, not signUp.

**Reply by sumchans** [reply]
Thanks @inian, just curious, if the user is new and need to signup with google, shouldn't we use the signup method as I have seen in some of the tutorials or does the signin method does both? here is one https://www.youtube.com/watch?v=_XM9ziOzWk4&t=194s. 

Also how do I get the user navigated to the sign in page when pressing sign up with google from the flutter app. For testing purposes I have put the signIn method call on the main method, but I see no navigation happening when the app loads. Would you be able to put some code samples please? Thanks again

**Reply by lokesh-coder** [reply]
Facing same issue. What I got to know is, internally [gotrue-dart](https://github.com/supabase/gotrue-dart) package just returns the [authorize](https://github.com/supabase/gotrue-dart#using) url for third party providers.

```dart
 final session =  await client.auth.signIn(provider: Provider.google);
 print(session.url); // returns https://<PROJECT_NAME>.supabase.co/auth/v1/authorize?provider=google
```
Once we got the URL, we need to manually trigger the browser tab to open this url. This can be done using [url_lancher](https://pub.dev/packages/url_launcher) or [flutter_web_auth](https://pub.dev/packages/flutter_web_auth) package. Right after the tab is opened and user completes the signIn/register, google redirects to callback url. 

This callback url will have access_token, and other necessary OAuth props in query params. Callback URL is taken from **Additional Redirect URLs** in Supabase Auth settings. But this has to be custom unique schema(other than http/s) so that it can be registered in mobile device and redirect to the app. We can set something like `myapp://callback`.

So the final callback url will looks like

```sh
myapp://callback#access_token=abc...xyz&expires_in=3600&provider_token=mno.efgh&refresh_token=jkl&token_type=bearer
```

from this we can parse the data and set the user session.

```dart
 final user = await client.auth.getSessionFromUrl(Uri.parse(callbackUrl));
```

To register the scheme `myapp://`, we need to set this, if `flutter_web_auth` is used.
in `android/app/src/main/AndroidManifest.xml`

```xml
<activity android:name="com.linusu.flutter_web_auth.CallbackActivity" >
    <intent-filter android:label="flutter_web_auth"  android:autoVerify="true">
        <action android:name="android.intent.action.VIEW" />
        <category android:name="android.intent.category.DEFAULT" />
        <category android:name="android.intent.category.BROWSABLE" />
        <data android:scheme="myapp" />
    </intent-filter>
</activity>
```

## Sample code

There might be a better solution, but this is what my temporary solution is.

```dart
import 'package:supabase/supabase.dart';
import 'package:flutter_web_auth/flutter_web_auth.dart';

  
final client = SupabaseClient('https://<PROJECT_NAME>.supabase.co','<SUPABASE_KEY>');

void signInWithGoogle(BuildContext context) async {
  final url = Uri.https('<PROJECT_NAME>.supabase.co', '/auth/v1/authorize', {
    'provider': 'google',
    'redirect_to': "myapp://callback",
  });
  final String callbackUri = await FlutterWebAuth.authenticate(
      url: url.toString(), callbackUrlScheme: "myapp");

  final session =
      await client.auth.getSessionFromUrl(Uri.parse(callbackUrl));
  print("user email => ${session.data?.user?.email}");
}
```

**Reply by phamhieu** [reply]
Thanks @lokesh-coder for the detail explanation. 

> internally gotrue-dart package just returns the authorize url for third party providers.

To keep the current library as a `dart` package. We can't use any specific flutter code to support features as launching url or session storage...
I'm open for a Supabase flutter package in the future that can help tackle these issues. We have a lot of work to do so we can't give firm deadlines about this.

**Reply by sumchans** [reply]
Thanks @lokesh-coder , this was nice and super detailed.

**Reply by netgfx** [reply]
Is there a typo? Should `callbackUri` be `callbackUrl` as it is used later on the `Uri.parse(callbackUrl)` ?

**Reply by clube-de-matematica** [reply]
How can I retrieve the callback URL in a flutter web app?
